bottlenose-dolphins/worldgate-track-trace
# [BDTT-55] Application Deployment & [BDTT-58] Dockerise Scraper Microservices

# Description
[BDTT-55: Application Deployment](https://bottlenose.atlassian.net/browse/BDTT-55)
[BDTT-58: Dockerise Scraper Microservices](https://bottlenose.atlassian.net/browse/BDTT-58)

List all acceptance criteria of user story and indicate which has been solved previously and which is solved currently

- [x] Dockerise KMTC and ONE scraper microservices
- [x] Solve Issues in DEV Docker env involving page crash
- [x] Create Terraform script to provision cloud infrastructure and deploy existing Images as containers using AWS' Elastic Container Service via Fargate 
- [x] Ensure in-application retrieval of microservice private IP addresses (/testRoute implemented in core_user microservice) 
- [ ]  BE-FE Integration within the hosted solution 

## Type of change

Please delete options that are not relevant.


Other noteworthy changes to bring to the attention of the PR Reviewer:

1. 2 different dockerfiles for each microservice --> this is because terraform does not facilitate the storing of `SRV` Records (i.e. IP and ports), only `A` records (i.e. ip), hence, 2 different dockerfiles were used - 1 to keep port numbering in the DEV docker env and the other to keep port 80 open in PROD Clod env. If we want to keep the ports, we could include an internal application load balancer, however, since an application load balancer does not have a fixed IP, this would require other components - can be solved by [implementing these](https://www.bluematador.com/blog/static-ips-for-aws-application-load-balancer#:~:text=One%20highly%20requested%20feature%20of,will%20in%20the%20near%20future.).

2. Dynamic retrieval of microservice Private IP addresses:
      -  Backend: python obtainIp.py 
      -  Frontend: user copy.js & user.js  (**Would appreciate feedback on this**)
<img width="927" alt="image" src="https://user-images.githubusercontent.com/46068652/218257221-17b7e3a1-9da5-4bb5-a39d-24b628f29f2e.png">
3. With reference to the diagram, I did not implement Public / Private Subnet Separation

# Checklist:

- [x] I have added the appropriate labels to my PR

# Steps to test Acceptance Criteria:

1. Follow the steps to deploy the solution infra and attach docker images onto infra using terraform
2. Container Hosted: By using the AWS UI to go into each task (container), we can retrieve the IP addresses of each container (ENI provided), all containers are mapped to port 80, and test queries, all of them should work. 
3. Inter-container communication: core_user service task has a route '/testRoute`: a call was made from the core_user microservice to CloudMap which returns the details of healthy scraper_ymlu microservices, the core_user then returns the private IP of 1 of these microservices, this can then be used to make an inter-container call (note: having a private IP means that we will not be able to make requests to the microservice from our computers - calls can be made from within the VPC, however, we can confirm that the provided private IP is accurate when using AWS UI)
